File: android/app/src/main/java/com/learningischange/tidbitstrivia/data/Entitlement.kt
package com.learningischange.tidbitstrivia.data

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.learningischange.tidbitstrivia.net.FirebaseNet

/**
 * Is this player a Tidbits Club member? The one gate every Club feature checks
 * (docs/CLUB-MONETIZATION-BUILD.md, MONETIZATION §7). Android twin of the Swift
 * EntitlementStore (TidbitsTrivia/Core/Networking/EntitlementStore.swift) and the web
 * mirror (js/entitlement.js) — same isClub gate, same fail-open discipline.
 *
 * Android has no local store yet (Class A = Google Play Billing, a later phase), so —
 * exactly like the web mirror — isClub is purely the REMOTE read: entitlements/{accountKey},
 * written by the Worker after a Merchant-of-Record purchase. Requires a verified-email
 * sign-in; the RTDB rule keys the read on emailOwners/{key} matching the auth token email.
 *
 * Fail OPEN: a transient read miss NEVER revokes Club (a paying member on a flaky
 * connection stays Club). Cache the last-known-good in SharedPreferences (mirror of
 * Duels.kt's init(ctx) pattern) so a returning member is Club instantly, before any
 * network round-trip.
 */
object Entitlement {
    private const val PREFS_NAME = "tidbits.entitlement"
    private const val KEY_IS_CLUB = "isClub"

    private var prefs: SharedPreferences? = null

    /** The gate. Seeded from the cached last-known-good so a returning member is Club
     *  instantly, before any network round-trip. */
    var isClub: Boolean by mutableStateOf(false); private set

    /** Called once from the Application so the cached Club flag survives process death —
     *  mirror of Duels.init(ctx). Must run before the first refresh()/read of isClub. */
    fun init(ctx: Context) {
        prefs = ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        isClub = prefs?.getBoolean(KEY_IS_CLUB, false) ?: false
    }

    /** True when this record grants an active Club membership. A subscription past
     *  `until` no longer grants; a lifetime (`until == null`) always does. */
    private fun grantsClub(ent: FirebaseNet.EntitlementRecord?): Boolean {
        if (ent == null || ent.tier != "club") return false
        val until = ent.until ?: return true
        return System.currentTimeMillis() < until
    }

    /** Recompute Club status. Safe to call at launch, after sign-in, and after a purchase. */
    suspend fun refresh() {
        if (!PlayerIdentity.signedIn || PlayerIdentity.profileId == null) {
            // Not signed in → no remote entitlement is readable. Don't aggressively revoke
            // a cached true (fail open); it re-confirms on the next signed-in refresh. A
            // fresh anon session with no cache is simply not Club.
            return
        }
        val key = PlayerIdentity.profileId ?: return
        try {
            val ent = FirebaseNet.loadEntitlement(key)
            set(grantsClub(ent))   // a clean read (incl. null -> not club) is authoritative
        } catch (e: Exception) {
            // transient RTDB error -> keep the cached answer (fail open)
        }
    }

    private fun set(club: Boolean) {
        isClub = club
        prefs?.edit()?.putBoolean(KEY_IS_CLUB, club)?.apply()
    }

    /** Sign-out clears the cached Club state (the next person on this device isn't you). */
    fun clearOnSignOut() { set(false) }
}

File: js/entitlement.js
// Tidbits Club membership on the web (docs/CLUB-MONETIZATION-BUILD.md, MONETIZATION §7).
// Mirror of the Swift EntitlementStore — same isClub gate, same fail-open discipline.
//
// The web has NO local store (Class A = StoreKit/Play/MSStore, which the web isn't), so on
// the web isClub is purely the REMOTE read: entitlements/{accountKey}, written by the
// Worker after a Merchant-of-Record purchase. Requires a verified-email sign-in — the RTDB
// rule keys the read on emailOwners/{key} === the token email.
//
// Fail OPEN: a transient read miss NEVER revokes Club (a paying member on a flaky
// connection stays Club). Cache the last-known-good so a returning member is Club instantly.

import { FirebaseNet } from './firebase.js';
import { Identity } from './identity.js';

const CACHE_KEY = 'tidbits.entitlement.isClub';

function grantsClub(ent) {
  if (!ent || ent.tier !== 'club') return false;
  if (ent.until == null) return true;            // lifetime / non-expiring
  return Date.now() < Number(ent.until);
}

export const Entitlement = {
  // Seed from the cached last-known-good so gating is correct before any network round-trip.
  isClub: (() => { try { return localStorage.getItem(CACHE_KEY) === '1'; } catch { return false; } })(),
  _subs: [],

  /** Subscribe to Club-status changes (fires immediately with the current value). */
  onChange(fn) { this._subs.push(fn); try { fn(this.isClub); } catch {} },
  _emit() { this._subs.forEach((f) => { try { f(this.isClub); } catch {} }); },

  _set(club) {
    const changed = this.isClub !== club;
    this.isClub = club;
    try { localStorage.setItem(CACHE_KEY, club ? '1' : '0'); } catch {}
    if (changed) this._emit();
  },

  /** Recompute Club status. Safe at launch, after sign-in, and after a purchase returns. */
  async refresh() {
    if (!Identity.signedIn || !Identity.profileId) {
      // Not signed in → no web entitlement is readable. Don't aggressively revoke a cached
      // true (fail open); it re-confirms on the next signed-in refresh. A fresh anon session
      // with no cache is simply not Club.
      return;
    }
    try {
      const ent = await FirebaseNet.loadEntitlement(Identity.profileId);
      this._set(grantsClub(ent));   // a clean read (incl. null → not club) is authoritative
    } catch {
      // transient error → keep cached (fail open)
    }
  },

  /** Sign-out clears Club (the next person on this device isn't you). */
  clearOnSignOut() { this._set(false); },
};
